stdlib/weak.ml

```diff
@@ -271,98 +271,45 @@ module Make (H : Hashtbl.HashedType) : (S with type data = H.t) = struct
     let h = H.hash d in
     add_aux t set (Some d) h (get_index t h)
 
+  (* General auxiliary function for searching for a particular value
+   * in a hash-set, and acting according to whether or not it's found *)
 
-  let find_or t d ifnotfound =
+  let find_aux ?(collect=false) t d found notfound =
```

Comment:
If `collect=true`, then `found` is unused; and if `collect=false`, the last argument passed to `notfound` is not used (it will always be empty).

I think it would be better to keep using a dedicated version for `find_all`, even if this means duplicating a few lines of code.

For the generic iterator (for find_opt, merge, find, remove, mem), I wonder what would be the performance impact of stopping with the CPS style, and return a sum type with the two constructors (each with an inline record). That's an extra allocation for the result, but fewer allocations for the closures (although some of them are constant), and perhaps a better behavior w.r.t. branch prediction.

Reply:
Yes. I disliked putting in the `collect` argument, it felt icky, and having refactored without it the code is nicer and the same speed (it surprised me that this didn't measurably speed up any of the operations).
I then tried the sum type approach you describe, but it added about 5% to everything. As you observe, most of the closures are constant (all of them, having broken `find_all` out, except the second one to `merge`).

Reply:
Do you mean, passing `t` and `b` back to the `notfound` continuation, to make it constant (perhaps one will need to lift its definition manually, because of the reference to `set` -- I'm not sure)?  I was going to suggest benchmarking that variant, to see if something is to be gained.

Reply:
I will benchmark that.